src/base/SanityChecker.ml

```diff
@@ -386,6 +389,83 @@ struct
       shadowing_libentries lmod.libs.lentries
   end
 
+  (* ********************************************** *)
+  (* ******** Check hashing builtins usage ******** *)
+  (* ********************************************** *)
+
+  module CheckHashingBuiltinsUsage = struct
+    let rec expr_iter (e, _annot) =
+      match e with
+      | Builtin ((builtin, _annot), _typ_args, args) -> (
+          match builtin with
+          | Builtin_sha256hash | Builtin_keccak256hash | Builtin_ripemd160hash
+            ->
+              forallM args ~f:(fun arg ->
+                  let type_of_arg = (ER.get_type (get_rep arg)).tp in
+                  match type_of_arg with
+                  | MapType _ | ADT _
+                  | PrimType (Msg_typ | Event_typ | Exception_typ) ->
```

Comment:
Will it work if the user applied hash in the body of a type function?

I'm thinking about something like:
```ocaml
let hash =
  tfun 'A =>
  fun (x: 'A) => keccak256hash x in
let hash_applied = @hash ByStr20 in
hash_applied
```

Reply:
OK, such cases cannot pass the typechecker.